### ISSUE TITLE
Inconsistent results/issues with cross browser use of LTI and xAPI and different LMS/VLEs

### ISSUE THREAD

**Issue by ronm123** [thexerteproject/xerteonlinetoolkits#831]
Hi all
I've been testing the LTI and xAPI functionality with the latest develop code after getting inconsistent results with 3.7.14. This was also the result of a provider testing with Blackboard and SCORM Cloud and reporting issues. Some of what isn't working with 3.7.14 is predictable due to required code not being part of that revision but the following is my results testing with the latest develop code.

With latest develop and my testing of scores going to the gradebook with LTI alone this is mostly working with Moodle but not with Blackboard. However there seems to be browser specific issues:

Moodle
Chrome passing details via moodle to lrs and Xerte and lti is passing score to Moodle gradebook
**Firefox** passing details via moodle to lrs and Xerte but lti **not passing score to Moodle gradebook**
Edge passing details via moodle to lrs and Xerte and lti is passing score to Moodle gradebook
IE passing details via moodle to lrs and Xerte and lti is passing score to Moodle gradebook
Opera passing details via moodle to lrs and Xerte and lti is passing score to Moodle gradebook

Blackboard
Chrome passing details via blackboard to lrs and Xerte and lti is passing score to BB gradebook**
Firefox passing details via blackboard to lrs and Xerte and lti is passing score to BB gradebook**
Edge passing details via blackboard to lrs and Xerte and lti is passing score to BB gradebook**
IE passing details via blackboard to lrs and Xerte but lti **not passing score to BB gradebook**
Opera passing details via blackboard to lrs and Xerte and lti is passing score to BB gradebook**

I've been testing with LTI only and a separate LO with LTI and xAPI. In moodle I created separate user accounts for each browser so that it's easy to see specific results but I couldn't easily do that in blackboard so my test LO contains a multiple choice question where the options are each browser name so that if the response works you see that in the dashboard.

Is anyone else able to test and replicate this?

@torinfo what's the best way to test why the LTI grades aren't working in Blackboard and aren't working with Firefox and Moodle?

**Comment by ronm123** [comment]
@torinfo following some further testing with latest develop I have edited my original post above and I'm a bit stuck now as to why there are different results with different browsers and LMS/VLEs. e.g. it would make sense if IE or Firefox didn't work regardless of LMS/VLE but this doesn't appear to be the case:

Current differences:
1. For some reason Firefox doesn't seem to send LTI scores to Moodle gradebook but Chrome, Edge, IE and Opera do.
2. However with Blackboard: Chrome, Firefox, Edge and Opera all send LTI scores to Blackboard but IE doesn't.
I/we have tested all of the above with LTI only, LTI and xAPI via Learning Locker and LTI and xAPI vis SCORM Cloud. The results are getting through to the LRS in each case but not the gradebook in the combinations described in 1. and 2.

What's the best way to try to troubleshoot each instance?

**Comment by torinfo** [comment]
@ronm123 
The Xerte code sends the LTI score during XTTerminate. It should not matter at all whether xAPI is enabled or not, or which LRS is being used.

There is a difference between LTI only (uses XTTerminate defined in modules/xerte/parent_templates/Nottingham/common_html5/js/xttracking_noop.js, line 1196) and LTI and xAPI (uses XTTerminate defined in modules/xerte/xAPI/xttracking_xapi.js, line 2908).

Also, when development is set to true, there should be a line in the debug log file coming from website_code/php/lti/sendgrade.php, line 40.

So as a start I would set a breakpoint in the debug console to one of the lines indicated above, and set development to true, and check the debug log file. 

I'll try to replicate the Moddle situation.

**Comment by ronm123** [comment]
@torinfo thanks Tom
I think the original testing I did with 3.7.14 was misleading because of different issues with xAPI but either way subsequent testing with latest develop is giving more consistent results but I'm not getting much further with the debugging etc.

There definitely seems to be a problem with Firefox and Moodle that doesn't exist with Firefox and Blackboard. However Firefox does seem to post to the Moodle gradebook with an earlier version e.g. Moodle 2.9.6 but not with Moodle 3.3.6 or Moodle 3.1.15+ and I've tested this with two different laptop/Firefox combinations.

With the debugging etc I was reminded that the tracking doesn't happen until the LO is closed and wondered if this is another on close event or something like that that isn't working with the Moodle + Firefox combination but is working with other browsers? I don't think the stop tracking page is liked up to the sending of LTI grades but wonder if it should be?

As I've said Firefox is sending grades with Blackboard and IE is sending grades with Moodle but IE isn't sending grades with Blackboard.

Obviously we need to resolve both if we can.

**Comment by torinfo** [comment]
@ronm123 

You can check if the grade is being sent by setting a breakpoint on the AJAX calls. But as Firefox is sending the grade to Blackboard, I would be surprised if the calls is not triggered. 

What has been logged in the debug file when you try to send a grade using firefox to Moodle?

The Stop tracking page is linked up with sending LTI grades, as the sending of LTI grades is done, when tracking is stopped.

**Comment by ronm123** [comment]
@torinfo as far as I can see when it works the Result Object has [grade] => 1 but when it doesn't work e.g. with Firefox the Result Object has just [grade] => 
I'll send you the debug.log for a Firefox attempt by direct email.

**Comment by ronm123** [comment]
@torinfo 
Ok pretty confident about this now but not sure exactly why the different browsers behave differently in different VLE/LMS – would make more sense if IE didn’t work in both BB and Moodle or FF didn’t work in both rather than each not working in the other. But anyway here’s what I think and can replicate…

By default in Moodle when adding an external tool the launch container is set to default which means the LO opens in the Moodle frame with a breadcrumb above it and when you get to the end of the LO you typically use the breadcrumb to return to the Moodle course homepage. For some reason this worked with sending to the gradebook with all browsers except Firefox.
If I change that launch container option to open in a new window then when closing the new window Firefox sends the grade correctly.

By default Blackboard was opening the LO in a new window but for some reason IE wasn’t sending to the gradebook when that windows was closed but all other browsers were.

Turning on Force tracking mode behaviour does indeed mean that the stop tracking page then shows the button and when I click the button the grade reaches the gradebook whether the LO itself is closed or not.

So basically turning on Force tracking mode fixes this for all browsers and both vles regardless of whether the LO is set to open in a new window or not. 

I wonder if there’s a way to fix each of the instances of closing the LO not sending the grade or whether there’s a better way to deal with this instead of an author having to enable that mode and also add the stop tracking page at the end?

Ron
